README.md
<h1 align="center">
  <img src="https://img.ideal-postcodes.co.uk/PAF%20API%20Logo@3x.png" alt="PAF API">
</h1>

> HTTP API that parses Postcode Address File records into correctly formatted address lines

![CI](https://github.com/ideal-postcodes/paf-api/workflows/CI/badge.svg)
![Release](https://github.com/ideal-postcodes/paf-api/workflows/Release/badge.svg)
[![Dependency Status](https://david-dm.org/ideal-postcodes/paf-api.png)](https://david-dm.org/ideal-postcodes/paf-api)
[![Coverage Status](https://coveralls.io/repos/github/ideal-postcodes/paf-api/badge.svg?branch=master)](https://coveralls.io/github/ideal-postcodes/paf-api?branch=master)
[![npm version](https://badge.fury.io/js/paf-api.svg)](https://badge.fury.io/js/paf-api)
[![install size](https://packagephobia.now.sh/badge?p=paf-api)](https://packagephobia.now.sh/result?p=paf-api)
[![](https://images.microbadger.com/badges/image/idealpostcodes/paf-api.svg)](https://microbadger.com/images/idealpostcodes/paf-api)

[![Deploy to Netlify](https://www.netlify.com/img/deploy/button.svg)](https://app.netlify.com/start/deploy?repository=https://github.com/ideal-postcodes/paf-api)

Parses Postcode Address File records into correctly formatted address recognised by Royal Mail according to its Clear Addressing Guidelines.

Produces consistent address lines, a post town line and a postcode line via HTTP

## Features

- Exposes [UK Clear Addressing](https://github.com/ideal-postcodes/uk-clear-addressing) as a HTTP service
- Deployable as a docker container or node.js application

![Correct Addressing](https://img.ideal-postcodes.co.uk/correct_address.gif)

## Links

- [Github Repository](https://github.com/ideal-postcodes/paf-api)
- [Test Live on paf-api.ideal-postcodes.dev](https://paf-api.ideal-postcodes.dev)
- [Docker Hub Image](https://hub.docker.com/r/idealpostcodes/paf-api)
- [UK Clear Addressing Project](https://github.com/ideal-postcodes/uk-clear-addressing)
- [NPM Package](https://www.npmjs.com/package/paf-api)
- [More information on Postcode Address File data attributes](https://ideal-postcodes.co.uk/documentation/paf-data)
- [PAF Programmer's Guide](https://js.ideal-postcodes.co.uk/guide.pdf)

## Getting Started

Try on [paf-api.ideal-postcodes.dev](https://paf-api.ideal-postcodes.dev)

```bash
curl -X POST \
  https://paf-api.ideal-postcodes.dev/parse \
  -H 'Content-Type: application/json' \
  -d '{ "sub_building_name": "Flat 8", "building_name": "Oxford House 110-114", "thoroughfare": "High Street" }'
```

### HTTP API Usage

This API responds to `POST` requests to `/parse` with a JSON payload representing a PAF Record

#### `POST /parse`

##### Request

```json
POST /parse

{
  "postcode": "WS11 5SB",
  "post_town": "CANNOCK",
  "thoroughfare": "Pye Green Road",
  "building_name": "Flower House 189A",
  "organisation_name": "S D Alcott Florists",
}
```

##### Response

```json
{
  "query": {
    "postcode": "WS11 5SB",
    "post_town": "CANNOCK",
    "thoroughfare": "Pye Green Road",
    "building_name": "Flower House 189A",
    "organisation_name": "S D Alcott Florists",
    "building_number": "",
    "sub_building_name": "",
    "dependant_locality": "",
    "double_dependant_locality": "",
    "dependant_thoroughfare": "",
    "po_box": "",
    "department_name": "",
  },
  "formatted": {
    "postcode": "WS11 5SB",
    "post_town": "CANNOCK",
    "line_1": "S D Alcott Florists",
    "line_2": "Flower House",
    "line_3": "189a Pye Green Road",
    "premise": "Flower House, 189a"
  }
}
```

#### Available Payload Attributes

```javascript
{
  building_number: string | number;
  building_name: string;
  sub_building_name: string;
  dependant_locality: string;
  double_dependant_locality: string;
  thoroughfare: string;
  dependant_thoroughfare: string;
  po_box: string;
  post_town: string;
  postcode: string;
  department_name: string;
  organisation_name: string;
}
```

### Deploy

#### Deploy as a docker container

Using the docker CLI

```
docker pull idealpostcodes/paf-api:latest

docker run -p 8080:8080 idealpostcodes/paf-api
```

Alternatively with docker-compose

```bash
git clone https://github.com/ideal-postcodes/paf-api && cd paf-api

docker-compose up -d
```

#### Deploy as a node.js process

Requires node.js 8 or above

```bash
git clone https://github.com/ideal-postcodes/paf-api && cd paf-api && npm install

npm start # Traffic served on 8080
```

#### NPM Module

The npm module exports an express app factory which takes a `pino` logger

```javascript
const { App } = require("paf-api");

const app = App({ logger: pino() });

http.createServer(app).listen(PORT);
```

#### Configuration

The following environment variables can be passed to the application.

```bash
PORT=8080 # Sets the port to listen on
```

## Licence

MIT



CHANGELOG.md
# [2.0.0](https://github.com/ideal-postcodes/paf-api/compare/1.0.0...2.0.0) (2020-07-22)


### chore

* **Node:** Drop explicit support for Node 8 ([db1dc6f](https://github.com/ideal-postcodes/paf-api/commit/db1dc6f))


### BREAKING CHANGES

* **Node:** Node 8 no longer forms part of CI testing

# [1.0.0](https://github.com/ideal-postcodes/paf-api/compare/0.1.0...1.0.0) (2020-06-25)


### chore

* **Node:** Drop node 8 support ([994b46f](https://github.com/ideal-postcodes/paf-api/commit/994b46f))


### BREAKING CHANGES

* **Node:** Drop support for Node 8

# [1.0.0](https://github.com/ideal-postcodes/paf-api/compare/0.1.0...1.0.0) (2020-06-25)


### chore

* **Node:** Drop node 8 support ([994b46f](https://github.com/ideal-postcodes/paf-api/commit/994b46f))


### BREAKING CHANGES

* **Node:** Drop support for Node 8

# [0.1.0](https://github.com/ideal-postcodes/paf-api/compare/0.0.1...0.1.0) (2019-06-06)


### Bug Fixes

* **Netlify:** Ensure dist dir available ([96c9064](https://github.com/ideal-postcodes/paf-api/commit/96c9064))


### Features

* **Semantic Release:** Automate releases ([6634dc4](https://github.com/ideal-postcodes/paf-api/commit/6634dc4))

# [0.1.0](https://github.com/ideal-postcodes/paf-api/compare/0.0.1...0.1.0) (2019-06-06)


### Bug Fixes

* **Netlify:** Ensure dist dir available ([96c9064](https://github.com/ideal-postcodes/paf-api/commit/96c9064))


### Features

* **Semantic Release:** Automate releases ([6634dc4](https://github.com/ideal-postcodes/paf-api/commit/6634dc4))

# Changelog

Any changes, including backwards incompatible changes will be listed here

## 0.0.1 (24/04/2019)

- Initial commit


test/app.ts
import request from "supertest";
import { postcodes, addresses } from "@ideal-postcodes/api-fixtures";
import { App } from "../lib/app";
import { assert } from "chai";
import P from "pino";
import { resolve } from "path";

const pino = P;
const destination = pino.destination(resolve(__dirname, "./test.log"));
const logger = pino(destination);
const config = { logger };
const app = App(config);
const OK = 200;
const NOT_FOUND = 404;

const testAddresses = [
  ...postcodes.success.body.result,
  ...addresses.success.body.result.hits,
];

describe("Application server", () => {
  it("returns 200 OK on /", async () => {
    await request(app)
      .get("/")
      .expect(OK);
  });

  it("provides readiness probe at /healthz", async () => {
    const response = await request(app)
      .get("/healthz")
      .expect(OK);
    assert.equal(response.body.status, "UP");
  });

  it("returns 404 on all other routes", async () => {
    await request(app)
      .get("/foo")
      .expect(NOT_FOUND);
  });

  describe("POST /parse", () => {
    for (const address of testAddresses) {
      const { line_1, postcode } = address;
      it(`returns a correctly parsed address for "${line_1}, ${postcode}"`, async () => {
        const addressQuery = {
          building_name: address.building_name,
          building_number: address.building_number,
          sub_building_name: address.sub_building_name,
          dependant_locality: address.dependant_locality,
          double_dependant_locality: address.double_dependant_locality,
          thoroughfare: address.thoroughfare,
          dependant_thoroughfare: address.dependant_thoroughfare,
          po_box: address.po_box,
          post_town: address.post_town,
          postcode: address.postcode,
          department_name: address.department_name,
          organisation_name: address.organisation_name,
        };

        const response = await request(app)
          .post("/parse")
          .send(addressQuery)
          .expect(OK);

        assert.deepEqual(response.body, {
          query: addressQuery,
          formatted: {
            line_1: address.line_1,
            line_2: address.line_2,
            line_3: address.line_3,
            post_town: address.post_town,
            postcode: address.postcode,
            premise: address.premise,
          },
        });
      });
    }

    it("handles empty requests", async () => {
      const addressQuery = {};
      const response = await request(app)
        .post("/parse")
        .send(addressQuery)
        .expect(OK);
      assert.deepEqual(response.body.formatted, {
        line_1: "",
        line_2: "",
        line_3: "",
        post_town: "",
        postcode: "",
        premise: "",
      });
    });
  });
});


test/index.ts
import { app } from "../lib/index";
import request from "supertest";

describe("Server", () => {
  it ("exports an application server", async () => {
    await request(app).get("/").expect(200); 
  });
});



test/netlify.ts
import { assert } from "chai";
import { handler, Event, Context } from "../lib/netlify";
import { postcodes, addresses } from "@ideal-postcodes/api-fixtures";

const OK = 200;
const BAD_REQUEST = 400;
const testAddresses = [
  ...postcodes.success.body.result,
  ...addresses.success.body.result.hits,
];

describe("Netlify FAAS", () => {
  describe("parse function", () => {
    for (const address of testAddresses) {
      const { line_1, postcode } = address;
      it(`returns a correctly parsed address for "${line_1}, ${postcode}"`, () => {
        const addressQuery = {
          building_name: address.building_name,
          building_number: address.building_number,
          sub_building_name: address.sub_building_name,
          dependant_locality: address.dependant_locality,
          double_dependant_locality: address.double_dependant_locality,
          thoroughfare: address.thoroughfare,
          dependant_thoroughfare: address.dependant_thoroughfare,
          po_box: address.po_box,
          post_town: address.post_town,
          postcode: address.postcode,
          department_name: address.department_name,
          organisation_name: address.organisation_name,
        };

        const event: Event = { body: JSON.stringify(addressQuery) };
        const context: Context = {};
        handler(event, context, (error, response) => {
          if (error) throw error;
          assert.equal(response.statusCode, OK);
          assert.deepEqual(JSON.parse(response.body), {
            query: addressQuery,
            formatted: {
              line_1: address.line_1,
              line_2: address.line_2,
              line_3: address.line_3,
              post_town: address.post_town,
              postcode: address.postcode,
              premise: address.premise,
            },
          });
        });
      });
    }
    it("returns 400 response on malformed payload", () => {
      const event: Event = { body: "}{" };
      const context: Context = {};

      handler(event, context, (error, response) => {
        assert.isNull(error);
        assert.equal(response.statusCode, BAD_REQUEST);
      });
    });
  });
});


netlify.sh
#! /usr/bin/env node

const marked = require("marked");
const { join } = require("path");
const { readFileSync, writeFileSync, existsSync, mkdirSync } = require("fs");

const path = "dist";
const file = "index.html";
const readme = readFileSync("./README.md", { encoding: "utf8" });

const html = `
<html>
  <head>
    <title>UK Clear Addressing API Usage</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/github-markdown-css/3.0.1/github-markdown.min.css" />
    <style>
      .markdown-body {
        box-sizing: border-box;
        min-width: 200px;
        max-width: 980px;
        margin: 0 auto;
        padding: 45px;
      }

      @media (max-width: 767px) {
        .markdown-body {
          padding: 15px;
        }
      }
    </style>
    <meta charset="UTF-8">
  </head>
  <body>
    <article class="markdown-body">
    ${marked(readme)}
    </article>
  </body>
</html>
`;

const outFile = join(__dirname, path, file);

// Create dist dir if it does not exist
existsSync(path) || mkdirSync(path);

// Write landing page to file
writeFileSync(outFile, html, { encoding: "utf8" });


lib/parse.ts
import { Address, AddressRecord } from "uk-clear-addressing";

interface FormattedAddress {
  premise: string;
  line_1: string;
  line_2: string;
  line_3: string;
  post_town: string;
  postcode: string;
}

interface AddressQuery {
  building_name: string;
  building_number: string;
  sub_building_name: string;
  dependant_locality: string;
  double_dependant_locality: string;
  thoroughfare: string;
  dependant_thoroughfare: string;
  po_box: string;
  post_town: string;
  postcode: string;
  department_name: string;
  organisation_name: string;
}

interface Response {
  query: AddressQuery;
  formatted: FormattedAddress;
}
export const parse = (body: AddressRecord): Response => {
  const {
    building_name = "",
    building_number = "",
    sub_building_name = "",
    dependant_locality = "",
    double_dependant_locality = "",
    thoroughfare = "",
    dependant_thoroughfare = "",
    po_box = "",
    post_town = "",
    postcode = "",
    department_name = "",
    organisation_name = "",
  } = body;

  const query = {
    building_name,
    building_number,
    sub_building_name,
    dependant_locality,
    double_dependant_locality,
    thoroughfare,
    dependant_thoroughfare,
    po_box,
    post_town,
    postcode,
    department_name,
    organisation_name,
  };

  const address = new Address(query);

  const formatted = {
    line_1: address.line_1,
    line_2: address.line_2,
    line_3: address.line_3,
    post_town: address.post_town,
    postcode: address.postcode,
    premise: address.premise,
  };

  return { query, formatted };
};


lib/netlify.ts
import { parse } from "./parse";

export interface Context {}

export interface Event {
  body: string;
}

interface Callback {
  (error: null | Error, response: Response): void;
}

type Headers = { [k: string]: string };

interface Response {
  statusCode: number;
  body: string;
  headers: Headers;
}

interface NetlifyEventHandler {
  (event: Event, context: Context, callback: Callback): void;
}

const OK = 200;
const BAD_REQUEST = 400;

interface ToJsonOptions {
  body: object;
  statusCode: number;
}

const toJson = ({ body, statusCode }: ToJsonOptions): Response => {
  const jsonString = JSON.stringify(body);
  return {
    body: jsonString,
    statusCode,
    headers: {
      "Content-Type": "application/json",
    },
  };
};

export const handler: NetlifyEventHandler = (event, _, callback) => {
  try {
    const query = JSON.parse(event.body);
    const body = parse(query);
    const statusCode = OK;
    callback(null, toJson({ statusCode, body }));
  } catch (_) {
    const statusCode = BAD_REQUEST;
    const body = { error: "Invalid JSON Submitted" };
    return callback(null, toJson({ statusCode, body }));
  }
};


lib/app.ts
import e from "express";
import P from "pino";
import expressPino from "express-pino-logger";
import { parse } from "./parse";

const express = e;
const OK = 200;
const NOT_FOUND = 404;

interface Config {
  logger: P.Logger;
}

export const App = (config: Config): Express.Application => {
  const { logger } = config;

  const app = express();

  app.use(expressPino({ logger }));
  app.use(express.json());

  app.get("/", (_, response) => {
    response.status(OK).send("<h1>Ping</h1>");
  });

  app.get("/healthz", (_, response) => {
    response.status(OK).json({ status: "UP" });
  });

  app.post("/parse", (request, response) => {
    response.status(OK).json(parse(request.body));
  });

  app.all("/", (_, response) => {
    response.status(NOT_FOUND).json({
      code: NOT_FOUND,
      message: "Resource not found. Try POST /parse",
    });
  });

  return app;
};
